activity/sidecar.go

```diff
@@ -0,0 +1,222 @@
+
+		//If an AWS credential lease is going to expire, assume all of them are and rewrite all of them
+		//All of our AWS credentials should have the same expiry, as STS expires them after an hour and
+		//we wrote all of them at the same time in the init task
+		if awsLease.Expiry.Before(threshold) {
+			if err := aws.WriteCredentials(currentConfig, vaultClient.Delegate); err != nil {
+				jww.FATAL.Fatalf("Could not write AWS credentials to replace expiring credentials: %v", err)
+			}
+			break
+		}
+	}
+
+	for _, sshLease := range leases.Current.SSHCertificates {
+		certificateFilename := filepath.Join(sshLease.OutputPath, sshsigning.SSHCertificate)
+		validBefore, err := sshsigning.ReadCertificateValidBefore(certificateFilename)
+
+		if err != nil {
+			jww.ERROR.Printf("could not get expiry date for SSH certificate %q: %v", certificateFilename, err)
+			continue
+		}
+
+		jww.DEBUG.Printf("SSH certificate %q is valid before %v", certificateFilename, validBefore)
+
+		// If the cert expires before "threshold" o'clock, then we should renew it. If renewing
+		// fails, then there are three outcomes. 1) fails because of permission denied, in which case we exit
+		// 2) fails because of another reason, but is still not expired (in which case we log, but keep going), or
+		// 3) fails because of another reason, but the ssh certificate is now invalid, in which case we exit.
+		if validBefore != uint64(ssh.CertTimeInfinity) && validBefore < uint64(threshold.Unix()) {
+			err := sshsigning.SignKey(vaultClient.Delegate, sshLease.OutputPath, sshLease.VaultMount, sshLease.VaultRole)
+
+			if errwrap.Contains(err, "Code: 403") {
+				jww.FATAL.Fatalf("Permission denied renewing SSH certificate %q.", certificateFilename)
+			} else {
+				if validBefore < uint64(time.Now().Unix()) {
+					jww.FATAL.Fatalf("Could not renew SSH lease for %q and it has expired: %v", certificateFilename, err)
+				}
+				jww.ERROR.Printf("Error renewing SSH lease for %q: %v", certificateFilename, err)
+			}
+		}
+	}
+}
+
+func performPeriodicSidecar(ctx context.Context, currentConfig cfg.Config, vaultClient vaultclient.VaultClient) {
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	signal.Notify(c, syscall.SIGTERM)
+
+	go func() {
+
+		jww.DEBUG.Printf("Performing auto renewal check on startup.")
+
+		renewLeases(ctx, currentConfig, vaultClient)
+
+		renewTicker := time.NewTicker(util.Flags.RenewInterval)
+		defer renewTicker.Stop()
+
+		jobCompletionTicker := time.NewTicker(18 * time.Second)
+		defer jobCompletionTicker.Stop()
+
+		var checkKubeAPITicks <-chan time.Time
+
+		jww.INFO.Printf("Lease renewal interval %v, completion check 10s", util.Flags.RenewInterval)
+		for {
+			select {
+			case <-ctx.Done():
+				jww.INFO.Printf("stopping renewal")
+				return
+			case <-renewTicker.C:
+				jww.INFO.Printf("Renewal Heartbeat")
+				renewLeases(ctx, currentConfig, vaultClient)
+			case <-checkKubeAPITicks:
```

Comment:
If we aren't implementing this we should leave it out for now

Reply:
If we are going to implement we should try and use constants for statuses. The k8s sdk probably has them already.